### PR TITLE
Improve log 'immediate' (former 'in place') mode and extend shell to support that mode

### PR DESCRIPTION
**Pull request by nordic-krch**
1. First change is fairly trivial: Added interrupt locking when doing in place log processing to ensure that one will not interrupt another. Description from commit message:

When CONFIG_LOG_IMMEDIATE is enabled then logs are processed in the caller context. In that case, it must be ensured that one log call will not be interrupted by another. As log can be called from any context, interrupt locking is the only option.

When CONFIG_LOG_IMMEDIATE  is used then logger backend also works in blocking way without interrupts so blocking them will not interfere with backend processing.

2. Second change is in shell and it adds in place support to shell. Description from commit message:

Extended shell to be able to process logs in place (in the context of a log call). In order to achieve that, shell was extended to  support for TX blocking operations. If CONFIG_LOG_IMMEDIATE  is enabled then shell instance attempts to be initialized in blocking TX mode. If fails to do so, shell log backend is disabled. If successfully enabled logs are processed and printed in the context of the call.

Due to that change, user may experience interleaved output as shell has no means to multiplex shell output with logger output. In extreme, huge amount of log messages may prevent shell thread execution and shell may become unresponsive.

### CODE REVIEW

File: subsys/shell/shell.c [hunk 1]
```diff
@@ -1175,7 +1175,7 @@ void shell_thread(void *shell_handle, void *arg_log_backend,
 				  &shell->ctx->signals[i]);
 	}
 
-	err = shell_start(shell);
+	err = shell->iface->api->enable(shell->iface, false);
```
Comment (pizi-nordic) [inline subsys/shell/shell.c]: What is the difference between this enable and start? Why these were separated?

File: subsys/shell/shell_log_backend.c [hunk 2]
```diff
@@ -18,10 +18,25 @@ int shell_log_backend_output_func(u8_t *data, size_t length, void *ctx)
...
-	backend->control_block->state = SHELL_LOG_BACKEND_ENABLED;
+	int err = 0;
+
+	if (IS_ENABLED(CONFIG_LOG_INPLACE_PROCESS)) {
+		const struct shell *shell;
+
+		shell = (const struct shell *)ctx;
+
+		/* Reenable transport in blocking mode */
+		err = shell->iface->api->enable(shell->iface, true);
```
Comment (pizi-nordic) [inline subsys/shell/shell_log_backend.c]: Could we have only one call to "enable" during startup?

File: subsys/shell/shell_log_backend.c [hunk 3]
```diff
@@ -158,9 +173,19 @@ static void put(const struct log_backend *const backend, struct log_msg *msg)
 		}
 
 		break;
-
+	case SHELL_LOG_BACKEND_BLOCKING:
+		/* Even though log message is handled notify shell thread which
+		 * will print command buffer after the log message.
+		 */
+		if (IS_ENABLED(CONFIG_MULTITHREADING)) {
```
Comment (pizi-nordic) [inline subsys/shell/shell_log_backend.c]: What if multithreading is disabled? We erased command prompt, how we can see it now?

File: subsys/shell/shell_uart.c [hunk 4]
```diff
@@ -155,18 +155,14 @@ static int uninit(const struct shell_transport *transport)
...
-	sh_uart->ctrl_blk->blocking = blocking;
+	sh_uart->ctrl_blk->blocking_tx = blocking_tx;
 
-	if (blocking) {
-		if (!IS_ENABLED(CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN)) {
-			k_timer_stop(sh_uart->timer);
-		}
+	if (blocking_tx) {
 #ifdef CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN
-		uart_irq_rx_disable(sh_uart->ctrl_blk->dev);
```
Comment (pizi-nordic) [inline subsys/shell/shell_uart.c]: Why this line is removed?

**Comment by pfalcon** [comment]
@nordic-krch :

> @pfalcon i know that it's late but treat it as my Christmas gift smile. 

Thanks, appreciated! ;-) It's still public holiday here today, will look into it tomorrow.

**Comment by nordic-krch** [comment]
@pfalcon thanks for checking.

I will take different approach as new requirement came out: mcuboot which wants in place logging with small footprint. Current approach is not small footprint as whole logger subsystem is still in use (~11k). I work toward in_place mode being direct string formatting and synchronous processing by active backends.

**Comment by pfalcon** [comment]
@nordic-krch, Sounds good, thanks.

**Comment by nordic-krch** [comment]
In place mode will be changed (#12406) and this PR will no long be valid. I'll come with new PR aligned with #12406 changes.

**Comment by pfalcon** [comment]
> For reference, ... is in _handle_obj_poll_events():

Hits again in https://github.com/zephyrproject-rtos/zephyr/pull/12465

**Comment by nordic-krch** [comment]
rebased on master where new approach to immediate logs has been merged. @pfalcon can you try that out?

File: subsys/shell/shell.c (hunk 1 above)
Comment (nordic-krch) [reply subsys/shell/shell.c]: shell_start has been moved not replaced. `enable` is part of transport interface

File: subsys/shell/shell_log_backend.c (hunk 2 above)
Comment (nordic-krch) [reply subsys/shell/shell_log_backend.c]: no, `enable` enables transport in given mode (blocking or nonblocking). Shell is by default started as asynchronous but can be reconfigured to blocking by the logger.

File: subsys/shell/shell_uart.c (hunk 4 above)
Comment (nordic-krch) [reply subsys/shell/shell_uart.c]: because this mode is now blocking TX. we only care about TX actually.

Comment (nordic-krch) [reply subsys/shell/shell_uart.c]: I've added comment to shell.h

**Comment by nordic-krch** [comment]
fairly gentle ping to all reviewers.

**Review by jakub-uC — APPROVED** [review]
We already discussed it, changes were applied.

**Comment by pfalcon** [comment]
> If
CONFIG_LOG_INPLACE_PROCESS is enabled then shell instance
attempts to be initialized in blocking TX mode. If fails to
do so, shell log backend is disabled.

And what "shell log backend is disabled" leads to? If to complete disabling of logging of output, then it's not what we want ;-). And if it leads to something else, e.g. switching to raw UART backend, then it rather should be mentioned explicitly in the commit message to not alarm users ;-).

**Review by pfalcon — APPROVED** [review]
So, I guess this another good step in the right direction, so definitely +1. Note that I finished writing that low-level driver, and no longer interact with logging issues on a daily basis. I need to get back to retesting all the previous issues, but so far that's backlogged.

I also remember that @carlescufi raised a question of renaming the config option for "immediate" mode, don't remember what was the decision. (Apparently, it happened in another ticket.)

**Comment by nordic-krch** [comment]
@pfalcon 

> And what "shell log backend is disabled" leads to? If to complete disabling of logging of output, then it's not what we want ;-).

This is the case if shell backend cannot work in blocking mode, like shell over bluetooth for example. In that case if user enables `LOG_IMMEDIATE` option that will lead to the situation where some log backends will no start. Only those which are capable of working synchronously, without interrupts will be enabled.

**Comment by nordic-krch** [comment]
> I also remember that @carlescufi raised a question of renaming the config option for "immediate" mode, don't remember what was the decision. (Apparently, it happened in another ticket.)

In place has been renamed to `LOG_IMMEDIATE`

**Comment by pfalcon** [comment]
Aha, I see, thanks!

**Review by carlescufi — APPROVED** [review]

**Pull request merged**